File: 2015/Day6/main.py
def process_lights(instructions):
    grid = [[False] * 1000 for _ in range(1000)]

    for instruction in instructions:
        parts = instruction.split()
        if parts[0] == 'turn':
            action = parts[1]
            start = tuple(map(int, parts[2].split(',')))
            end = tuple(map(int, parts[4].split(',')))
            for i in range(start[0], end[0] + 1):
                for j in range(start[1], end[1] + 1):
                    grid[i][j] = (action == 'on')
        elif parts[0] == 'toggle':
            start = tuple(map(int, parts[1].split(',')))
            end = tuple(map(int, parts[3].split(',')))
            for i in range(start[0], end[0] + 1):
                for j in range(start[1], end[1] + 1):
                    grid[i][j] = not grid[i][j]
    
    return sum(sum(row) for row in grid)


def process_lights_brightness(instructions):
    grid = [[0] * 1000 for _ in range(1000)]

    for instruction in instructions:
        parts = instruction.split()
        if parts[0] == 'turn':
            action = parts[1]
            start = tuple(map(int, parts[2].split(',')))
            end = tuple(map(int, parts[4].split(',')))
            for i in range(start[0], end[0] + 1):
                for j in range(start[1], end[1] + 1):
                    if action == 'on':
                        grid[i][j] += 1
                    elif action == 'off':
                        grid[i][j] = max(0, grid[i][j] - 1)
        elif parts[0] == 'toggle':
            start = tuple(map(int, parts[1].split(',')))
            end = tuple(map(int, parts[3].split(',')))
            for i in range(start[0], end[0] + 1):
                for j in range(start[1], end[1] + 1):
                    grid[i][j] += 2
        
    return sum(sum(row) for row in grid)

with open('input.txt', 'r') as file:
    instructions = file.read().splitlines()

print(f"Part 1: {process_lights(instructions)}")
print(f"Part 2: {process_lights_brightness(instructions)}")

File: 2015/Day6/main.dart
import 'dart:io';

int processLights(List<String> instructions) {
  var grid = List.generate(1000, (_) => List.filled(1000, false));

  for (var instruction in instructions) {
    List<String> parts = instruction.split(' ');
    if (parts[0] == 'turn') {
      String action = parts[1];
      List<int> start = parts[2].split(',').map(int.parse).toList();
      List<int> end = parts[4].split(',').map(int.parse).toList();
      for (int i = start[0]; i <= end[0]; i++) {
        for (int j = start[1]; j <= end[1]; j++) {
          grid[i][j] = (action == 'on');
        }
      }
    } else if (parts[0] == 'toggle') {
      List<int> start = parts[1].split(',').map(int.parse).toList();
      List<int> end = parts[3].split(',').map(int.parse).toList();
      for (int i = start[0]; i <= end[0]; i++) {
        for (int j = start[1]; j <= end[1]; j++) {
          grid[i][j] = !grid[i][j];
        }
      }
    }
  }

  return grid.fold(0, (sum, row) => sum + row.where((light) => light).length);
}

int processLightBrightness(List<String> instructions) {
  var grid = List.generate(1000, (_) => List.filled(1000, 0));

  for (var instruction in instructions) {
    List<String> parts = instruction.split(' ');
    if (parts[0] == 'turn') {
      var action = parts[1];
      List<int> start = parts[2].split(',').map(int.parse).toList();
      List<int> end = parts[4].split(',').map(int.parse).toList();
      for (int i = start[0]; i <= end[0]; i++) {
        for (int j = start[1]; j <= end[1]; j++) {
          if (action == 'on') {
            grid[i][j]++;
          } else if (action == 'off') {
            grid[i][j] = (grid[i][j] - 1).clamp(0, grid[i][j]);
          }
        }
      }
    } else if (parts[0] == 'toggle') {
      List<int> start = parts[1].split(',').map(int.parse).toList();
      List<int> end = parts[3].split(',').map(int.parse).toList();
      for (int i = start[0]; i <= end[0]; i++) {
        for (int j = start[1]; j <= end[1]; j++) {
          grid[i][j] += 2;
        }
      }
    }
  }

  return grid.fold(0, (sum, row) => sum + row.reduce((a, b) => a + b));
}

void main() async {
  var instructions = await File('input.txt').readAsLines();

  print('Part 1: ${processLights(instructions)}');
  print('Part 2: ${processLightBrightness(instructions)}');
}
